### ISSUE TITLE
[FEATURE REQUEST] Change Title/Message in a Dialog

### ISSUE THREAD

**Issue by AlbinoGeek** [fyne-io/fyne#1566]
### Is your feature request related to a problem? Please describe:
I have some long-running interactions that use `ProgressInfiniteDialog` and it would be great to simply be able to change the message shown as progress continued, instead of hiding the dialog and creating a completely new one every time a new stage of the process begins.

### Is it possible to construct a solution with the existing API?
No

### Describe the solution you'd like to see:

Consider indeterminate action with multiple steps, such as a program installation or download+extract.

```golang
total, progress, err := StartDownload("...")

// assume "total" was 0/-1/undefined for this example (e.g: webserver didn't tell us)
dlg := dialog.NewProgressInfinite("Downloading Update", "Starting download...", w)
select {
  case n, done := <-progress:
    if done { break }
    dlg.SetMessage(fmt.Sprintf("%d bytes retrieved...", n))
}

dlg.SetTitle("Extracting Update")
dlg.SetMessage("Starting file extraction...")
select {
  case f, done := <- events:
    if done { break }
    dlg.SetMessage(fmt.Sprintf("Extracting %s...", f)
}
```

**Comment by andydotxyz** [comment]
Have you considered using a regular progress dialog seeing as you are presenting progress?

**Comment by AlbinoGeek** [comment]
I still cannot `SetTitle` or `SetMessage` a Progress Dialog, only `SetValue`. Each step is independent, and each may or may not have progress (e.g: the download has a change for progress) -- but I'd still change the text (if possible) of the dialog to show that.

Of course, I suppose a custom dialog could be used for this (with `widget.NewProgressBarInfinite()`) -- just a question of how much `Dialog` is meant to be able to do as shortcuts.

**Comment by andydotxyz** [comment]
Apologies, I thought we had exposed this feature through the progress dialog: `widget.NewProgressBar().TextFormatter`.
Maybe that offers some help?

I am keepipng this issue open for implementation - just wanted to find you a solution in the meantime.

**Comment by AlbinoGeek** [comment]
More specifically, the issue here was that `ProgressDialog` (and well, none of the dialogs) expose their `Message` label for you to call `SetText` on after creating them, which then means I have to keep making new dialogs instead of re-using existing ones (which, coupled with the other bugs about dialogs atm, makes them harder to use.)

https://pkg.go.dev/fyne.io/fyne@v1.4.0/dialog#NewProgress

As can be seen in the docs (code wasn't linking properly) -- there's no public way to access the title of the dialog, nor the message displayed -- nor any methods to shortcut to those things `SetTitle`, `SetMessage`, etc. `DialogProgress` may not be the best example, but other dialogs have the same concern.

---

Other bugs:
- Sometimes dialogs don't show up when called (issues opened by others)
- If a dialog is currently being shown and you try to show a new one, <unknown behaviour> ensues.

Unknown Behaviour examples:
- Sometimes the first dialog stays up, and the second `Show` is ignored entirely.
- Sometimes the first dialog closes, and the second dialog `Show`s (expected result)
- Sometimes the first dialog closes, and nothing else happens (???)

I have not nailed down why this happens, hence no issue opened specifically for it yet.

**Comment by andydotxyz** [comment]
Are you definitely calling `Show()` on each of the dialogs, not just `New...Dialog`? There are issues in this area we are fixing in 2.0 that could not be back-ported as the fix would have broken behaviour.
I /think/ the list of other / unknown issues above could all be tracked to that.

**Comment by AlbinoGeek** [comment]
Correct! I will be making a repository to test/show off this issue shortly.

Basic premise:

1) Create and Show ProgressInfinite (Checking for Updates...)
1) Hide 1, Create and Show Progress (Downloading Update...)
1) Hide 2, Create and Show ProgressInfinite (Extracting Update...)
1) Hide 3, Create and Show Information (or Error) (Result)

I have to explicitly save handles and hide the previous dialogs, or the unknown behaviour happens.

Of course, I could replace this whole process with a custom dialog, or modal (when those work with wrapped labels, such as a long filename, see other issues) -- but I figured this could be done with the dialog shortcuts.

Does the `dialog` package save a reference to the last dialog, for dismissing or otherwise? Such as the case when `ShowXXXDialog` is used instead of `dlg := NewXXXDialog(...); dlg.Show(); dlg.Hide()`

**Comment by andydotxyz** [comment]
> I have to explicitly save handles and hide the previous dialogs, or the unknown behaviour happens.

Yes, the developer is responsible for dismissing dialogs if the user does not take a dismissal action.

> Does the dialog package save a reference to the last dialog, for dismissing or otherwise?

The only reference saved is the one you get from NewXxxDialog... otherwise the object tree knows it is present to draw it, but will not make any assumptions about closing - the dialog will be hidden by the developer when Hide() is called, or by the user when they dismiss the dialog with OK/Cancel.

**Comment by AlbinoGeek** [comment]
> Yes, the developer is responsible for dismissing dialogs if the user does not take a dismissal action.

What is the expected behaviour in this case then?

1) Developer shows a Dialog (say... `ShowInformation(...)`)
2) While dialog is still shown, Developer shows another Dialog (say... `ShowError(...)`)

And, should this expected behaviour be documented?

**Comment by andydotxyz** [comment]
Newly opened dialogs will appear over whatever content is currently visible.
I don’t know if this needs to be documented, but if it does where do you think would be suitable?
